### PR TITLE
web: add a client-side logstore

### PR DESCRIPTION
**Pull request by nicks**
Hello @jazzdan, @hyu,

Please review the following commits I made in branch nicks/clientlogs:

c8bf2935da179bfa38f0f7dae2564f4ffcf2e7ec (2019-12-06 14:31:52 -0500)
web: add a client-side logstore
currently, this is a straight port of the server-side logstore

over time, we expect this logstore to have much richer methods
for filtering logs by manifest, log level, etc

### CODE REVIEW

File: web/src/LogStore.ts [hunk 1]
```diff
@@ -0,0 +1,191 @@
+// Client-side log store, which helps client side rendering and filtering of logs.
+//
+// Looslely adapted from the data structures in
```
Comment (jazzdan) [inline web/src/LogStore.ts]:
```suggestion
// Loosely adapted from the data structures in
```

File: web/src/LogStore.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+class LogSegment {
+  spanId: string
+  time: string
+  text: string
+  continuesLine: boolean
+
+  constructor(seg: Proto.webviewLogSegment) {
+    this.spanId = seg.spanId || ""
```
Comment (jazzdan) [inline web/src/LogStore.ts]: We can use `??` here now
```suggestion
    this.spanId = seg.spanId ?? ""
```

ditto on the ones below

File: web/src/LogStore.ts [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  spans: { [key: string]: LogSpan }
+  segments: LogSegment[]
+
+  constructor() {
+    this.spans = {}
+    this.segments = []
+  }
+
+  append(logList: Proto.webviewLogList) {
+    let newSpans = logList.spans as { [key: string]: Proto.webviewLogSpan }
```
Comment (jazzdan) [inline web/src/LogStore.ts]: Why do we cast this instead of handling the case where `logList.spans` isn't defined?

File: web/src/LogStore.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        }
+      }
+    }
+
+    let segments = logList.segments ?? []
+    segments.forEach(segment => {
+      let newSegment = new LogSegment(segment)
+      let span = this.spans[newSegment.spanId]
+      if (!span) {
+        // If we don't have the span for this log, we can't meaninfully print it,
```
Comment (jazzdan) [inline web/src/LogStore.ts]:
```suggestion
        // If we don't have the span for this log, we can't meaningfully print it,
```

File: web/src/LogStore.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    })
+  }
+
+  allLog() {
+    return this.manifestLog("")
+  }
+
+  manifestLog(mn: string) {
+    let lastLineCompleted = false
+    let allLogs = mn == ""
```
Comment (jazzdan) [inline web/src/LogStore.ts]: should we use `===` here and on the line below?

File: web/src/LogStore.ts (hunk 1 above)
Comment (nicks) [reply web/src/LogStore.ts]: done

File: web/src/LogStore.ts (hunk 2 above)
Comment (nicks) [reply web/src/LogStore.ts]: done

File: web/src/LogStore.ts (hunk 3 above)
Comment (nicks) [reply web/src/LogStore.ts]: logList.spans is always defined. this isn't about runtime behavior. this is about compile-time behavior. The type definition is "object" and needs to be cast to the correct type. i don't know why the typescript definitions use "object"

File: web/src/LogStore.ts (hunk 4 above)
Comment (nicks) [reply web/src/LogStore.ts]: done

File: web/src/LogStore.ts (hunk 5 above)
Comment (nicks) [reply web/src/LogStore.ts]: done

**Pull request merged**